lib/wibox/widget/slider.lua

```diff
@@ -0,0 +1,473 @@
+-- @param color
+
+--- The handle shape.
+-- @beautiful beautiful.slider_handle_shape
+-- @tparam[opt=gears shape rectangle] gears.shape shape
+-- @see gears.shape
+
+--- The bar (background) shape.
+-- @beautiful beautiful.slider_bar_shape
+-- @tparam[opt=gears shape rectangle] gears.shape shape
+-- @see gears.shape
+
+--- The bar (background) height.
+-- @beautiful beautiful.slider_bar_height
+-- @param number
+
+--- The bar (background) margins.
+-- @beautiful beautiful.slider_bar_margins
+-- @tparam[opt={}] table margins
+-- @tparam[opt=0] number margins.left
+-- @tparam[opt=0] number margins.right
+-- @tparam[opt=0] number margins.top
+-- @tparam[opt=0] number margins.bottom
+
+--- The slider handle margins.
+-- @beautiful beautiful.slider_handle_margins
+-- @tparam[opt={}] table margins
+-- @tparam[opt=0] number margins.left
+-- @tparam[opt=0] number margins.right
+-- @tparam[opt=0] number margins.top
+-- @tparam[opt=0] number margins.bottom
+
+--- The bar (background) color.
+-- @beautiful beautiful.slider_bar_color
+-- @param color
+
+local properties = {
+    -- Handle
+    handle_shape         = shape.rectangle,
+    handle_color         = false,
+    handle_margins       = {},
+    handle_width         = false,
+    handle_border_width  = 0,
+    handle_border_color  = false,
+
+    -- Bar
+    bar_shape            = shape.rectangle,
+    bar_height           = false,
+    bar_color            = false,
+    bar_margins          = {},
+    bar_border_width     = 0,
+    bar_border_color     = false,
+
+    -- Content
+    value                = 0,
+    minimum              = 0,
+    maximum              = 100,
+}
+
+-- Create the accessors
+for prop in pairs(properties) do
+    slider["set_"..prop] = function(self, value)
+        local changed = self._private[prop] ~= value
+        self._private[prop] = value
+
+        if changed then
+            self:emit_signal("property::"..prop)
+            self:emit_signal("widget::redraw_needed")
+        end
+    end
+
+    slider["get_"..prop] = function(self)
+        -- Ignoring the false's is on purpose
+        return self._private[prop] == nil
+            and properties[prop]
+            or self._private[prop]
+    end
+end
+
+-- Add some validation to set_value
+function slider:set_value(value)
+    value = math.min(value, self:get_maximum())
+    value = math.max(value, self:get_minimum())
+    local changed = self._private.value ~= value
+
+    self._private.value = value
+
+    if changed then
+        self:emit_signal( "property::value"       )
+        self:emit_signal( "widget::redraw_needed" )
+    end
+end
+
+local function get_extremums(self)
+    local min = self._private.minimum or properties.minimum
+    local max = self._private.maximum or properties.maximum
+    local interval = max - min
+
+    return min, max, interval
+end
+
+local function get_value(self, width, x)
+    local _, _, interval = get_extremums(self)
+
+    return math.floor((x*interval)/width)
```

Comment:
I'd suggest inlining `get_value()` into its only caller `move_handle`. Also, why the `math.floor`? This is (by default) a percentage value that you are rounding, so something between 0 and 100. Having this as an integer shouldn't make any difference unless the slider is exactly 100 pixel wide. I suggest getting rid of `math.floor`.